File: backend/api/utils/type_guards.py
"""
Type Validators Module for Astrology Data

This module provides type validation functions for astrology data types.
It implements runtime type checking with descriptive error messages.
"""

import json
from dataclasses import dataclass
from typing import (
    Any,
    Dict,
    List,
    Literal,
    Optional,
    Protocol,
    TypedDict,
    TypeGuard,
    TypeVar,
    Union,
    cast,
)

# Type alias for unknown types in runtime validation
Unknown = Union[None, bool, int, float, str, List[Any], Dict[str, Any]]

# Type variables for generic type checking
T = TypeVar("T")

# Literal type for discriminating between astrology data types
AstrologyDataType = Literal[
    "AstrologyChart", "UserProfile", "NumerologyData", "Unknown"
]


# TypedDict definitions for astrology data structures
class Planet(TypedDict, total=False):
    """Type definition for Planet objects"""

    name: str  # Could be restricted to planet names
    sign: str  # Could be restricted to zodiac signs
    degree: float
    position: float
    house: str
    retrograde: Optional[bool]
    aspects: Optional[List[Dict[str, Unknown]]]


class House(TypedDict):
    """Type definition for House objects"""

    house: int
    number: int
    sign: str  # Could be restricted to zodiac signs
    degree: float
    cusp: float
    ruler: str  # Could be restricted to planet names


class Aspect(TypedDict):
    """Type definition for Aspect objects"""

    planet1: str  # Could be restricted to planet names
    planet2: str  # Could be restricted to planet names
    type: str  # Could be restricted to aspect types
    orb: float
    applying: str


class Asteroid(TypedDict):
    """Type definition for Asteroid objects"""

    name: str
    sign: str  # Could be restricted to zodiac signs
    degree: float
    house: str


class Angle(TypedDict):
    """Type definition for Angle objects"""

    name: str  # Could be restricted to angle names
    sign: str  # Could be restricted to zodiac signs
    degree: float
    position: float


# Protocol class for chart-like objects (structural typing)
class ChartLike(Protocol):
    """Protocol for any chart-like object with celestial bodies"""

    planets: List[Planet]
    houses: List[House]
    aspects: List[Aspect]


class BirthData(TypedDict):
    """Type definition for birth data"""

    date: str
    time: str
    location: str


class UserProfile(TypedDict, total=False):
    """Type definition for user profile"""

    userId: str
    birthData: BirthData
    preferences: Optional[Dict[str, Unknown]]


class NumerologyData(TypedDict):
    """Type definition for numerology data"""

    lifePath: int
    destiny: int
    personalYear: int


class AstrologyChart(TypedDict, total=False):
    """Type definition for astrology chart"""

    planets: List[Planet]
    houses: List[House]
    aspects: List[Aspect]
    asteroids: List[Asteroid]
    angles: List[Angle]
    name: Optional[str]
    date: Optional[str]
    userId: Optional[str]


def is_planet(value: object) -> TypeGuard[Planet]:
    """Type guard for Planet objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    return (
        isinstance(obj.get("name"), str)
        and isinstance(obj.get("sign"), str)
        and isinstance(obj.get("degree"), (int, float))
        and isinstance(obj.get("position"), (int, float))
        and isinstance(obj.get("house"), str)
        and (
            obj.get("retrograde") is None
            or isinstance(obj.get("retrograde"), bool)
        )
        and (
            obj.get("aspects") is None or isinstance(obj.get("aspects"), list)
        )
    )


def is_house(value: object) -> TypeGuard[House]:
    """Type guard for House objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    return (
        isinstance(obj.get("house"), int)
        and isinstance(obj.get("number"), int)
        and isinstance(obj.get("sign"), str)
        and isinstance(obj.get("degree"), (int, float))
        and isinstance(obj.get("cusp"), (int, float))
        and isinstance(obj.get("ruler"), str)
    )


def is_aspect(value: object) -> TypeGuard[Aspect]:
    """Type guard for Aspect objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    return (
        isinstance(obj.get("planet1"), str)
        and isinstance(obj.get("planet2"), str)
        and isinstance(obj.get("type"), str)
        and isinstance(obj.get("orb"), (int, float))
        and isinstance(obj.get("applying"), str)
    )


def is_asteroid(value: object) -> TypeGuard[Asteroid]:
    """Type guard for Asteroid objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    return (
        isinstance(obj.get("name"), str)
        and isinstance(obj.get("sign"), str)
        and isinstance(obj.get("degree"), (int, float))
        and isinstance(obj.get("house"), str)
    )


def is_angle(value: object) -> TypeGuard[Angle]:
    """Type guard for Angle objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    return (
        isinstance(obj.get("name"), str)
        and isinstance(obj.get("sign"), str)
        and isinstance(obj.get("degree"), (int, float))
        and isinstance(obj.get("position"), (int, float))
    )


def is_astrology_chart(value: object) -> TypeGuard[AstrologyChart]:
    """Type guard for AstrologyChart objects with deep validation"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    # Check for required arrays
    if not all(
        isinstance(obj.get(key), list)
        for key in ["planets", "houses", "aspects", "asteroids", "angles"]
    ):
        return False

    # Validate each array item with safe type checking
    planets = obj.get("planets", [])
    houses = obj.get("houses", [])
    aspects = obj.get("aspects", [])
    asteroids = obj.get("asteroids", [])
    angles = obj.get("angles", [])
    
    return (
        isinstance(planets, list) and all(is_planet(p) for p in planets)
        and isinstance(houses, list) and all(is_house(h) for h in houses)
        and isinstance(aspects, list) and all(is_aspect(a) for a in aspects)
        and isinstance(asteroids, list) and all(is_asteroid(a) for a in asteroids)
        and isinstance(angles, list) and all(is_angle(a) for a in angles)
    )


def is_user_profile(value: object) -> TypeGuard[UserProfile]:
    """Type guard for UserProfile objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    if not isinstance(obj.get("userId"), str):
        return False

    birth_data = obj.get("birthData")
    if not isinstance(birth_data, dict):
        return False

    # Cast birth_data for type checking
    birth_dict = cast(Dict[str, Unknown], birth_data)

    return (
        isinstance(birth_dict.get("date"), str)
        and isinstance(birth_dict.get("time"), str)
        and isinstance(birth_dict.get("location"), str)
    )


def is_numerology_data(value: object) -> TypeGuard[NumerologyData]:
    """Type guard for NumerologyData objects"""
    if not isinstance(value, dict):
        return False

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], value)

    return (
        isinstance(obj.get("lifePath"), int)
        and isinstance(obj.get("destiny"), int)
        and isinstance(obj.get("personalYear"), int)
    )


def get_astrology_data_type(data: object) -> str:
    """
    Type-safe data discriminator
    Returns the specific type name of the astrology-related data
    """
    if is_astrology_chart(data):
        return "AstrologyChart"
    elif is_user_profile(data):
        return "UserProfile"
    elif is_numerology_data(data):
        return "NumerologyData"
    else:
        return "Unknown"


def validate_astrology_chart(chart: object) -> List[str]:
    """
    Validates an astrology chart structure and reports specific validation errors  # noqa: E501
    Returns an array of validation errors, empty if valid
    """
    errors: List[str] = []

    if not isinstance(chart, dict):
        return ["Chart must be a dictionary"]

    # Cast to dict for type checking
    obj = cast(Dict[str, Unknown], chart)

    # Check required properties
    planets = obj.get("planets")
    if not isinstance(planets, list):
        errors.append("Chart is missing planets array")
    elif len(planets) == 0:
        errors.append("Chart must have at least one planet")
    else:
        # Validate each planet
        for i, planet in enumerate(planets):
            if not is_planet(planet):
                errors.append(f"Invalid planet at index {i}")

    houses = obj.get("houses")
    if not isinstance(houses, list):
        errors.append("Chart is missing houses array")
    elif len(houses) != 12:
        errors.append("Chart must have exactly 12 houses")
    else:
        # Validate each house
        for i, house in enumerate(houses):
            if not is_house(house):
                errors.append(f"Invalid house at index {i}")

    aspects = obj.get("aspects")
    if not isinstance(aspects, list):
        errors.append("Chart is missing aspects array")
    else:
        # Validate each aspect
        for i, aspect in enumerate(aspects):
            if not is_aspect(aspect):
                errors.append(f"Invalid aspect at index {i}")

    asteroids = obj.get("asteroids")
    if not isinstance(asteroids, list):
        errors.append("Chart is missing asteroids array")
    else:
        # Validate each asteroid
        for i, asteroid in enumerate(asteroids):
            if not is_asteroid(asteroid):
                errors.append(f"Invalid asteroid at index {i}")

    angles = obj.get("angles")
    if not isinstance(angles, list):
        errors.append("Chart is missing angles array")
    else:
        # Validate each angle
        for i, angle in enumerate(angles):
            if not is_angle(angle):
                errors.append(f"Invalid angle at index {i}")

    return errors


@dataclass
class ChartValidationResult:
    """Result of chart validation and parsing"""

    chart: Optional[AstrologyChart]
    errors: List[str]
    is_valid: bool


def safe_parse_astrology_chart(json_string: str) -> ChartValidationResult:
    """
    Safely attempts to parse a JSON string into an AstrologyChart
    Returns a ChartValidationResult with the parsed chart (or None if invalid) and any validation errors  # noqa: E501
    """
    try:
        parsed = json.loads(json_string)
        validation_errors = validate_astrology_chart(parsed)

        if validation_errors:
            return ChartValidationResult(
                chart=None, errors=validation_errors, is_valid=False
            )

        # We've validated that it's an AstrologyChart
        return ChartValidationResult(
            chart=cast(AstrologyChart, parsed), errors=[], is_valid=True
        )
    except json.JSONDecodeError as e:
        return ChartValidationResult(
            chart=None, errors=[str(e)], is_valid=False
        )
    except Exception as e:
        return ChartValidationResult(
            chart=None,
            errors=[f"Unknown parsing error: {str(e)}"],
            is_valid=False,
        )


def is_astrology_data(value: object) -> bool:
    """
    Type guard to check if the value is any valid astrology data type
    """
    return (
        is_astrology_chart(value)
        or is_user_profile(value)
        or is_numerology_data(value)
    )

File: packages/types/src/type-guards.ts
/**
 * Type Guards for Astrology Data Types
 * 
 * This module provides type predicates (type guards) for safely narrowing types
 * at runtime. These guards allow for more precise type checking in TypeScript
 * and help prevent runtime errors by validating data structures.
 */

import type {
  AstrologyChart,
  Planet,
  House,
  Aspect,
  Asteroid,
  ChartAngles,
  UserProfile,
  NumerologyData,
} from './astrology.types';

/**
 * Type guard for Planet objects (compatible with bridge system)
 */
export function isPlanet(value: unknown): value is Planet {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  return (
    typeof obj['name'] === 'string' &&
    typeof obj['sign'] === 'string' &&
    typeof obj['position'] === 'number' &&
    typeof obj['house'] === 'number' &&
    (obj['degree'] === undefined || typeof obj['degree'] === 'number') &&
    (obj['retrograde'] === undefined || typeof obj['retrograde'] === 'boolean') &&
    (obj['speed'] === undefined || typeof obj['speed'] === 'number') &&
    (obj['dignity'] === undefined ||
      ['domicile', 'exaltation', 'fall', 'detriment'].includes(obj['dignity'] as string)) &&
    (obj['aspects'] === undefined || Array.isArray(obj['aspects']))
  );
}

/**
 * Type guard for House objects (compatible with bridge system)
 */
export function isHouse(value: unknown): value is House {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  return (
    typeof obj['number'] === 'number' &&
    typeof obj['cusp'] === 'number' &&
    typeof obj['sign'] === 'string' &&
    typeof obj['ruler'] === 'string' &&
    typeof obj['degree'] === 'number' &&
    typeof obj['size'] === 'number' &&
    (obj['modern_ruler'] === undefined || typeof obj['modern_ruler'] === 'string') &&
    (obj['contains_planets'] === undefined || Array.isArray(obj['contains_planets']))
  );
}

/**
 * Type guard for Aspect objects (compatible with bridge system)
 */
export function isAspect(value: unknown): value is Aspect {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  return (
    typeof obj['aspect_type'] === 'string' &&
    typeof obj['planet1'] === 'string' &&
    typeof obj['planet2'] === 'string' &&
    typeof obj['orb'] === 'number' &&
    typeof obj['applying'] === 'boolean' &&
    typeof obj['exact'] === 'boolean' &&
    typeof obj['power'] === 'number' &&
    typeof obj['aspect_angle'] === 'number' &&
    (obj['separating'] === undefined || typeof obj['separating'] === 'boolean') &&
    (obj['mutual_reception'] === undefined || typeof obj['mutual_reception'] === 'boolean')
  );
}

/**
 * Type guard for Asteroid objects (compatible with bridge system)
 */
export function isAsteroid(value: unknown): value is Asteroid {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  return (
    typeof obj['name'] === 'string' &&
    typeof obj['position'] === 'number' &&
    typeof obj['degree'] === 'number' &&
    typeof obj['sign'] === 'string' &&
    typeof obj['house'] === 'number' &&
    typeof obj['retrograde'] === 'boolean' &&
    typeof obj['speed'] === 'number'
  );
}

/**
 * Type guard for ChartAngles objects (compatible with bridge system)
 */
export function isAngle(value: unknown): value is ChartAngles {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  return (
    (obj['ascendant'] === undefined || typeof obj['ascendant'] === 'number') &&
    (obj['descendant'] === undefined || typeof obj['descendant'] === 'number') &&
    (obj['midheaven'] === undefined || typeof obj['midheaven'] === 'number') &&
    (obj['imumcoeli'] === undefined || typeof obj['imumcoeli'] === 'number')
  );
}

/**
 * Type guard for AstrologyChart objects
 * Performs deep validation of nested structures
 */
export function isAstrologyChart(value: unknown): value is AstrologyChart {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  // Check for required structures
  if (
    typeof obj['planets'] !== 'object' || obj['planets'] === null ||
    !Array.isArray(obj['houses']) ||
    !Array.isArray(obj['aspects']) ||
    (obj['asteroids'] !== undefined && (typeof obj['asteroids'] !== 'object' || obj['asteroids'] === null)) ||
    typeof obj['angles'] !== 'object' || obj['angles'] === null
  ) {
    return false;
  }

  // Validate planets record
  const planets = obj['planets'] as Record<string, unknown>;
  for (const planetData of Object.values(planets)) {
    if (!isPlanet(planetData)) return false;
  }

  // Validate each house
  if (!obj['houses'].every(isHouse)) return false;

  // Validate each aspect
  if (!obj['aspects'].every(isAspect)) return false;

  // Validate asteroids if present
  if (obj['asteroids']) {
    const asteroids = obj['asteroids'] as Record<string, unknown>;
    for (const asteroidData of Object.values(asteroids)) {
      if (!isAsteroid(asteroidData)) return false;
    }
  }

  // Validate angles object
  if (!isAngle(obj['angles'])) return false;

  return true;
}

/**
 * Type guard for UserProfile objects
 */
export function isUserProfile(value: unknown): value is UserProfile {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  if (typeof obj['userId'] !== 'string') {
    return false;
  }

  // Check birthData structure
  if (
    obj['birthData'] === null ||
    obj['birthData'] === undefined ||
    typeof obj['birthData'] !== 'object'
  ) {
    return false;
  }

  const birthData = obj['birthData'] as Record<string, unknown>;

  return (
    typeof birthData['date'] === 'string' &&
    typeof birthData['time'] === 'string' &&
    typeof birthData['location'] === 'string'
  );
}

/**
 * Type guard for NumerologyData objects
 */
export function isNumerologyData(value: unknown): value is NumerologyData {
  if (value === null || value === undefined || typeof value !== 'object') {
    return false;
  }

  const obj = value as Record<string, unknown>;

  return (
    typeof obj['lifePath'] === 'number' &&
    typeof obj['destiny'] === 'number' &&
    typeof obj['personalYear'] === 'number'
  );
}

/**
 * Type-safe data discriminator
 * Returns the specific type name of the astrology-related data
 */
export function getAstrologyDataType(
  data: unknown
): 'AstrologyChart' | 'UserProfile' | 'NumerologyData' | 'Unknown' {
  if (isAstrologyChart(data)) return 'AstrologyChart';
  if (isUserProfile(data)) return 'UserProfile';
  if (isNumerologyData(data)) return 'NumerologyData';
  return 'Unknown';
}

/**
 * Type guard to check if the value is any valid astrology data type
 */
export function isAstrologyData(
  value: unknown
): value is AstrologyChart | UserProfile | NumerologyData {
  return (
    isAstrologyChart(value) || isUserProfile(value) || isNumerologyData(value)
  );
}

/**
 * Validates an astrology chart structure and reports specific validation errors
 * @returns An array of validation errors, empty if valid
 */
export function validateAstrologyChart(chart: unknown): string[] {
  const errors: string[] = [];

  if (chart === null || chart === undefined || typeof chart !== 'object') {
    return ['Chart must be an object'];
  }

  const obj = chart as Record<string, unknown>;

  // Check required properties
  if (typeof obj['planets'] !== 'object' || obj['planets'] === null) {
    errors.push('Chart is missing planets record');
  } else {
    const planets = obj['planets'] as Record<string, unknown>;
    if (Object.keys(planets).length === 0) {
      errors.push('Chart must have at least one planet');
    } else {
      // Validate each planet
      Object.entries(planets).forEach(([key, planet]) => {
        if (!isPlanet(planet)) {
          errors.push(`Invalid planet at key ${key}`);
        }
      });
    }
  }

  if (!Array.isArray(obj['houses'])) {
    errors.push('Chart is missing houses array');
  } else {
    // Validate each house
    obj['houses'].forEach((house, index) => {
      if (!isHouse(house)) {
        errors.push(`Invalid house at index ${index}`);
      }
    });
  }

  if (!Array.isArray(obj['aspects'])) {
    errors.push('Chart is missing aspects array');
  } else {
    // Validate each aspect
    obj['aspects'].forEach((aspect, index) => {
      if (!isAspect(aspect)) {
        errors.push(`Invalid aspect at index ${index}`);
      }
    });
  }

  if (obj['asteroids'] !== undefined) {
    if (typeof obj['asteroids'] !== 'object' || obj['asteroids'] === null) {
      errors.push('Asteroids must be a record if present');
    } else {
      const asteroids = obj['asteroids'] as Record<string, unknown>;
      Object.entries(asteroids).forEach(([key, asteroid]) => {
        if (!isAsteroid(asteroid)) {
          errors.push(`Invalid asteroid at key ${key}`);
        }
      });
    }
  }

  if (typeof obj['angles'] !== 'object' || obj['angles'] === null) {
    errors.push('Chart is missing angles object');
  } else {
    if (!isAngle(obj['angles'])) {
      errors.push('Invalid angles object');
    }
  }

  return errors;
}

/**
 * Safely attempts to parse a JSON string into an AstrologyChart
 * @returns A tuple with the parsed chart (or null if invalid) and any validation errors
 */
export function safeParseAstrologyChart(
  jsonString: string
): [AstrologyChart | null, string[]] {
  try {
    const parsed = JSON.parse(jsonString) as Record<string, unknown>;
    const validationErrors = validateAstrologyChart(parsed);

    if (validationErrors.length > 0) {
      return [null, validationErrors];
    }

    // Only cast to AstrologyChart if it has passed validation
    if (isAstrologyChart(parsed)) {
      return [parsed, []];
    }

    return [null, ['Invalid chart data structure']];
  } catch (error) {
    return [
      null,
      [error instanceof Error ? error.message : 'Unknown parsing error'],
    ];
  }
}
